File: ABC/abc345/d/d.py
from sys import stdin,setrecursionlimit
setrecursionlimit(10 ** 6)
import pypyjit
pypyjit.set_param('max_unroll_recursion=-1')

def myin():
    return stdin.readline().rstrip()

def myin_sp():
    return stdin.readline().rstrip().split()

def myin_sp_i():
    return list(map(int,myin_sp()))

def myin_sp_s():
    return list(map(str,myin_sp()))

N=H=W=0
tills = []
def solve(board,unused,cur):
    while cur<H*W and board[cur]>0:
        cur+=1
    if cur==H*W:
        return True
    ny = cur//W
    nx = cur%W
    for i,(a,b) in enumerate(tills):
        if (unused&(1<<i))>0:
            if 0<=ny+a-1<H and 0<=nx+b-1<W:
                flag = True
                for dy in range(a):
                    for dx in range(b):
                        if board[(ny+dy)*W+nx+dx]>0:
                            flag = False
                            break
                    if not flag:
                        break
                if flag:
                    for dy in range(a):
                        for dx in range(b):
                            board[(ny+dy)*W+nx+dx]=1
                    f2 = solve(board,unused^(1<<i),cur)
                    if f2:
                        return True
                    for dy in range(a):
                        for dx in range(b):
                            board[(ny+dy)*W+nx+dx]=0
            if a!=b and 0<=ny+b-1<H and 0<=nx+a-1<W:
                flag = True
                for dy in range(b):
                    for dx in range(a):
                        if board[(ny+dy)*W+nx+dx]>0:
                            flag = False
                            break
                    if not flag:
                        break
                if flag:
                    for dy in range(b):
                        for dx in range(a):
                            board[(ny+dy)*W+nx+dx]=1
                    f3 = solve(board,unused^(1<<i),cur)
                    if f3:
                        return True
                    for dy in range(b):
                        for dx in range(a):
                            board[(ny+dy)*W+nx+dx]=0
    return False
                
            
def main():
    global N,H,W,tills
    N,H,W = myin_sp_i()
    tills = []
    for i in range(N):
        tills.append(myin_sp_i())
    if(solve([0 for _ in range(H*W)],(1<<N)-1,0)):
        print("Yes")
    else:
        print("No")

if __name__ == "__main__":
    main()

File: ABC/abc345/d/d.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
int N, H, W;
vector<pair<int, int>> tills;

int solve(int unused, vector<vector<int>> board, int prev_y)
{
    int is_full = 1;
    for (int i = prev_y; i < H; i++)
    {
        for (int j = 0; j < W; j++)
        {
            if (board[i][j] == 0)
            {
                is_full = 0;
                break;
            }
        }
        if (!is_full)
            break;
        prev_y = i + 1;
    }
    if (is_full)
    {
        return 1;
    }
    for (int n = 0; n < N; n++)
    {
        if (unused & (1 << n))
        {
            int A = tills[n].first;
            int B = tills[n].second;
            for (int y = prev_y; y < H - A + 1; y++)
            {
                for (int x = 0; x < W - B + 1; x++)
                {
                    if (board[y][x] > 0)
                        continue;
                    int can = 1;
                    for (int dy = 0; dy < A; dy++)
                    {
                        for (int dx = 0; dx < B; dx++)
                        {
                            if (board[y + dy][x + dx] > 0)
                            {
                                can = 0;
                                break;
                            }
                        }
                    }
                    if (can)
                    {
                        for (int dy = 0; dy < A; dy++)
                        {
                            for (int dx = 0; dx < B; dx++)
                            {
                                board[y + dy][x + dx] = 1;
                            }
                        }
                        if (solve(unused ^ (1 << n), board, prev_y))
                        {
                            return 1;
                        }
                        for (int dy = 0; dy < A; dy++)
                        {
                            for (int dx = 0; dx < B; dx++)
                            {
                                board[y + dy][x + dx] = 0;
                            }
                        }
                    }
                }
            }
            if (A != B)
            {
                for (int y = prev_y; y < H - B + 1; y++)
                {
                    for (int x = 0; x < W - A + 1; x++)
                    {
                        if (board[y][x] > 0)
                            continue;
                        int can = 1;
                        for (int dy = 0; dy < B; dy++)
                        {
                            for (int dx = 0; dx < A; dx++)
                            {
                                if (board[y + dy][x + dx] > 0)
                                {
                                    can = 0;
                                    break;
                                }
                            }
                        }
                        if (can)
                        {
                            for (int dy = 0; dy < B; dy++)
                            {
                                for (int dx = 0; dx < A; dx++)
                                {
                                    board[y + dy][x + dx] = 1;
                                }
                            }
                            if (solve(unused ^ (1 << n), board, prev_y))
                            {
                                return 1;
                            }
                            for (int dy = 0; dy < B; dy++)
                            {
                                for (int dx = 0; dx < A; dx++)
                                {
                                    board[y + dy][x + dx] = 0;
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    return 0;
}
int main()
{
    cin >> N >> H >> W;
    for (int i = 0; i < N; i++)
    {
        int A, B;
        cin >> A >> B;
        tills.push_back({A, B});
    }
    vector<vector<int>> board(H, vector<int>(W, 0));
    if (solve((1 << N) - 1, board, 0))
    {
        cout << "Yes\n";
    }
    else
    {
        cout << "No\n";
    }
    return 0;
}